### ISSUE TITLE
torch.topk is non-deterministic on ROCm for bfloat16/float16 when the reduced dim exceeds 256

### ISSUE THREAD

**Issue by stefanskiasan** [pytorch/pytorch#196177]
### 🐛 Describe the bug

On ROCm, `torch.topk` returns **different indices across identical invocations**
for `bfloat16` and `float16` inputs once the reduced dimension exceeds 256.
`float32` is unaffected. This is not the documented tie-breaking ambiguity — the
majority of affected rows have **strictly distinct** values at the k-boundary.

#### Reproducer

```python
import torch
torch.manual_seed(0)

for E in (64, 128, 192, 256, 257, 384, 512, 1024, 2048):
    x = torch.randn(256, E, device="cuda", dtype=torch.bfloat16)
    K = 8
    ref = torch.topk(x, K, dim=-1).indices
    changed = torch.zeros(256, dtype=torch.bool, device="cuda")
    for _ in range(10):
        changed |= (torch.topk(x, K, dim=-1).indices != ref).any(dim=-1)
    # rows whose k-th and (k+1)-th values differ -> selection is unambiguous
    w = torch.topk(x.float(), K + 1, dim=-1).values
    distinct = w[:, K-1] != w[:, K]
    print(f"{E:>5} cols: {changed.sum().item():>3}/256 rows changed, "
          f"{(changed & distinct).sum().item():>3} of them with distinct values")
```

#### Result

```
   64 cols:   0/256 rows changed,   0 of them with distinct values
  128 cols:   0/256 rows changed,   0 of them with distinct values
  192 cols:   0/256 rows changed,   0 of them with distinct values
  256 cols:   0/256 rows changed,   0 of them with distinct values
  257 cols:  93/256 rows changed,  77 of them with distinct values
  384 cols:  55/256 rows changed,  46 of them with distinct values
  512 cols:  72/256 rows changed,  55 of them with distinct values
 1024 cols:  58/256 rows changed,  45 of them with distinct values
 2048 cols:  35/256 rows changed,  26 of them with distinct values
```

The threshold is exactly at **256 → 257**, which suggests a switch between a
single-block path and a multi-pass/radix path, with the latter not being
deterministic.

By dtype, 257 columns, k=8, 10 invocations of the same tensor:

| dtype | distinct selections | rows differing (of 256) |
|---|---:|---:|
| bfloat16 | 10 / 10 | up to 62 |
| float16 | 10 / 10 | up to 10 |
| float32 | 1 / 10 | 0 |

#### Why it matters

MoE routers select top-k experts from the gating output. A router that calls
`torch.topk` on the gating dtype — commonly bfloat16 — inherits this, and a
different expert is not a rounding difference but a structurally different
result. Expert counts above 256 are common (GLM-5.3 has 257, DeepSeek-V3 has
257), which places them just past the threshold.

Workaround that works: cast to `float32` before the `topk` call. The cost is
negligible at these sizes.

### Versions

```
torch      : 2.12.0+git6bbd260
ROCm       : 7.2.3
GPU        : AMD Instinct MI355X (gfx950)
OS         : Ubuntu 24.04
```

Not tested on CUDA — the reproducer above should show whether it is
ROCm-specific.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @mruberry @kurtamohler

**Comment by naromero77amd** [comment]
@stefanskiasan This is an older version of PyTorch and an older version of ROCm. Please re-test with either the recent stable PyTorch 2.14 wheel or the nightly PyTorch wheel.

**Comment by malfet** [comment]
bfloat16 have only 65K distinct values, so I'm not surprised it happens to have duplicate in tensors larger than 64K elements, and then results of topk are non deterministic.

Though neither CUDA nor MPS exhibit the same failure

**Comment by naromero77amd** [comment]
@stefanskiasan ROCm and NVIDIA's topk should have identical behavior with respect to non-determinism. Are you seeing different behavior on NVIDIA?
